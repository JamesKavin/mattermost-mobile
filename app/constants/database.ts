// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import keyMirror from '@utils/key_mirror';

export const MM_TABLES = {
    APP: {
        INFO: 'Info',
        GLOBAL: 'Global',
        SERVERS: 'Servers',
    },
    SERVER: {
        CATEGORY: 'Category',
        CATEGORY_CHANNEL: 'CategoryChannel',
        CHANNEL: 'Channel',
        CHANNEL_INFO: 'ChannelInfo',
        CHANNEL_MEMBERSHIP: 'ChannelMembership',
        CONFIG: 'Config',
        CUSTOM_EMOJI: 'CustomEmoji',
        DRAFT: 'Draft',
        FILE: 'File',
        GROUP: 'Group',
        GROUP_CHANNEL: 'GroupChannel',
        GROUP_MEMBERSHIP: 'GroupMembership',
        GROUP_TEAM: 'GroupTeam',
        MY_CHANNEL: 'MyChannel',
        MY_CHANNEL_SETTINGS: 'MyChannelSettings',
        MY_TEAM: 'MyTeam',
        POST: 'Post',
        POSTS_IN_CHANNEL: 'PostsInChannel',
        POSTS_IN_THREAD: 'PostsInThread',
        PREFERENCE: 'Preference',
        REACTION: 'Reaction',
        ROLE: 'Role',
        SYSTEM: 'System',
        TEAM: 'Team',
        TEAM_CHANNEL_HISTORY: 'TeamChannelHistory',
        TEAM_MEMBERSHIP: 'TeamMembership',
        TEAM_SEARCH_HISTORY: 'TeamSearchHistory',
        THREAD: 'Thread',
        THREADS_IN_TEAM: 'ThreadsInTeam',
        THREAD_PARTICIPANT: 'ThreadParticipant',
        TEAM_THREADS_SYNC: 'TeamThreadsSync',
        USER: 'User',
    },
};

export const MIGRATION_EVENTS = keyMirror({
    MIGRATION_ERROR: null,
    MIGRATION_STARTED: null,
    MIGRATION_SUCCESS: null,
});

export const SYSTEM_IDENTIFIERS = {
    CONFIG: 'config',
    CURRENT_CHANNEL_ID: 'currentChannelId',
    LAST_UNREAD_CHANNEL_ID: 'lastUnreadChannelId',
    CURRENT_TEAM_ID: 'currentTeamId',
    CURRENT_USER_ID: 'currentUserId',
    DATA_RETENTION_POLICIES: 'dataRetentionPolicies',
    EXPANDED_LINKS: 'expandedLinks',
    GLOBAL_THREADS_TAB: 'globalThreadsTab',
    LAST_DISMISSED_BANNER: 'lastDismissedBanner',
    LAST_SERVER_VERSION_CHECK: 'LastServerVersionCheck',
    LICENSE: 'license',
    ONLY_UNREADS: 'onlyUnreads',
    PUSH_VERIFICATION_STATUS: 'pushVerificationStatus',
    RECENT_CUSTOM_STATUS: 'recentCustomStatus',
    RECENT_MENTIONS: 'recentMentions',
    RECENT_REACTIONS: 'recentReactions',
    SESSION_EXPIRATION: 'sessionExpiration',
    TEAM_HISTORY: 'teamHistory',
    WEBSOCKET: 'WebSocket',
};

export const GLOBAL_IDENTIFIERS = {
    DEVICE_TOKEN: 'deviceToken',
    DONT_ASK_FOR_REVIEW: 'dontAskForReview',
    FIRST_LAUNCH: 'firstLaunch',
    LAST_ASK_FOR_REVIEW: 'lastAskForReview',
    MULTI_SERVER_TUTORIAL: 'multiServerTutorial',
    PROFILE_LONG_PRESS_TUTORIAL: 'profileLongPressTutorial',
    ONBOARDING: 'onboarding',
};

export enum OperationType {
    CREATE = 'CREATE',
    UPDATE = 'UPDATE',
    DELETE = 'DELETE',
  }

// The only two types of databases in the app
export enum DatabaseType {
    DEFAULT,
    SERVER,
  }

export default {
    GLOBAL_IDENTIFIERS,
    MM_TABLES,
    MIGRATION_EVENTS,
    SYSTEM_IDENTIFIERS,
};
