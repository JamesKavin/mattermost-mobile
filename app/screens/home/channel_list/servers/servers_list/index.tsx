// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useCallback} from 'react';
import {useIntl} from 'react-intl';
import {ListRenderItemInfo, StyleSheet, View} from 'react-native';
import {FlatList} from 'react-native-gesture-handler';

import {useServerUrl} from '@context/server';
import {useTheme} from '@context/theme';
import {useIsTablet} from '@hooks/device';
import BottomSheetContent from '@screens/bottom_sheet/content';
import {addNewServer} from '@utils/server';

import ServerItem from './server_item';

import type ServersModel from '@typings/database/models/app/servers';

type Props = {
    servers: ServersModel[];
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    contentContainer: {
        marginVertical: 4,
    },
});

const keyExtractor = (item: ServersModel) => item.url;

const ServerList = ({servers}: Props) => {
    const intl = useIntl();
    const isTablet = useIsTablet();
    const serverUrl = useServerUrl();
    const theme = useTheme();

    const onAddServer = useCallback(async () => {
        addNewServer(theme);
    }, [servers]);

    const renderServer = useCallback(({item: t, index}: ListRenderItemInfo<ServersModel>) => {
        return (
            <ServerItem
                highlight={index === 0}
                isActive={t.url === serverUrl}
                server={t}
            />
        );
    }, []);

    return (
        <BottomSheetContent
            buttonIcon='plus'
            buttonText={intl.formatMessage({id: 'servers.create_button', defaultMessage: 'Add a server'})}
            onPress={onAddServer}
            showButton={true}
            showTitle={!isTablet}
            testID='server_list'
            title={intl.formatMessage({id: 'your.servers', defaultMessage: 'Your servers'})}
        >
            <View style={styles.container}>
                <FlatList
                    data={servers}
                    renderItem={renderServer}
                    keyExtractor={keyExtractor}
                    contentContainerStyle={styles.contentContainer}
                />
            </View>
        </BottomSheetContent>
    );
};

export default ServerList;
