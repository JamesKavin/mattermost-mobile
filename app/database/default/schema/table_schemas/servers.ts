// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {tableSchema} from '@nozbe/watermelondb';

import {MM_TABLES} from '@constants/database';

const {SERVERS} = MM_TABLES.DEFAULT;

export default tableSchema({
    name: SERVERS,
    columns: [
        {name: 'db_path', type: 'string'},
        {name: 'display_name', type: 'string'},
        {name: 'mention_count', type: 'number'},
        {name: 'unread_count', type: 'number'},
        {name: 'url', type: 'string', isIndexed: true},
    ],
});
