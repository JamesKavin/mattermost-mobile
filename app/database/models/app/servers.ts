// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Model} from '@nozbe/watermelondb';
import {field} from '@nozbe/watermelondb/decorators';

import {MM_TABLES} from '@constants/database';

const {SERVERS} = MM_TABLES.APP;

/**
 * The Server model will help us to identify the various servers a user will log in; in the context of
 * multi-server support system.  The db_path field will hold the App-Groups file-path
 */
export default class Servers extends Model {
    /** table (name) : servers */
    static table = SERVERS;

    /** db_path : The file path where the database is stored */
    @field('db_path') dbPath!: string;

    /** display_name : The server display name */
    @field('display_name') displayName!: string;

    /** mention_count : The number of mention on this server */
    @field('mention_count') mentionCount!: number;

    /** unread_count : The number of unread messages on this server */
    @field('unread_count') unreadCount!: number;

    /** url : The online address for the Mattermost server */
    @field('url') url!: string;

    /** last_active_at: The last time this server was active */
    @field('last_active_at') lastActiveAt!: number;

    /** is_secured: Determines if the protocol used for this server url is HTTP or HTTPS */
    @field('is_secured') isSecured!: boolean;
}
