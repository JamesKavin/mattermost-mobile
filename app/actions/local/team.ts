// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DatabaseManager from '@database/manager';
import {prepareDeleteTeam, queryMyTeamById, removeTeamFromTeamHistory} from '@queries/servers/team';

import type TeamModel from '@typings/database/models/servers/team';

export const removeUserFromTeam = async (serverUrl: string, teamId: string) => {
    const serverDatabase = DatabaseManager.serverDatabases[serverUrl];
    if (!serverDatabase) {
        return;
    }

    const {operator, database} = serverDatabase;

    const myTeam = await queryMyTeamById(database, teamId);
    if (myTeam) {
        const team = await myTeam.team.fetch() as TeamModel;
        const models = await prepareDeleteTeam(team);
        const system = await removeTeamFromTeamHistory(operator, team.id, true);
        if (system) {
            models.push(...system);
        }
        if (models.length) {
            try {
                await operator.batchRecords(models);
            } catch {
                // eslint-disable-next-line no-console
                console.log('FAILED TO BATCH CHANGES FOR REMOVE USER FROM TEAM');
            }
        }
    }
};
