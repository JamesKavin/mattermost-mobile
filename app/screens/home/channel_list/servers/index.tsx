// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useCallback, useEffect, useRef, useState} from 'react';
import {IntlShape, useIntl} from 'react-intl';
import {StyleSheet} from 'react-native';

import ServerIcon from '@components/server_icon';
import {useServerUrl} from '@context/server';
import {useTheme} from '@context/theme';
import {subscribeAllServers} from '@database/subscription/servers';
import {subscribeUnreadAndMentionsByServer} from '@database/subscription/unreads';
import {useIsTablet} from '@hooks/device';
import {bottomSheet} from '@screens/navigation';

import ServerList from './servers_list';

import type ServersModel from '@typings/database/models/app/servers';
import type MyChannelModel from '@typings/database/models/servers/my_channel';
import type {UnreadMessages, UnreadSubscription} from '@typings/database/subscriptions';

const subscriptions: Map<string, UnreadSubscription> = new Map();

const styles = StyleSheet.create({
    icon: {
        alignItems: 'center',
        justifyContent: 'center',
        position: 'absolute',
        zIndex: 10,
        top: 10,
        left: 16,
        width: 40,
        height: 40,
    },
});

const sortServers = (servers: ServersModel[], intl: IntlShape) => {
    function serverName(s: ServersModel) {
        if (s.displayName === s.url) {
            return intl.formatMessage({id: 'servers.default', defaultMessage: 'Default Server'});
        }

        return s.displayName;
    }

    return servers.sort((a, b) => {
        return serverName(a).localeCompare(serverName(b));
    });
};

export default function Servers() {
    const intl = useIntl();
    const [total, setTotal] = useState<UnreadMessages>({mentions: 0, unread: false});
    const registeredServers = useRef<ServersModel[]|undefined>();
    const currentServerUrl = useServerUrl();
    const isTablet = useIsTablet();
    const theme = useTheme();

    const updateTotal = () => {
        let unread = false;
        let mentions = 0;
        subscriptions.forEach((value) => {
            unread = unread || value.unread;
            mentions += value.mentions;
        });
        setTotal({mentions, unread});
    };

    const unreadsSubscription = (serverUrl: string, myChannels: MyChannelModel[]) => {
        const unreads = subscriptions.get(serverUrl);
        if (unreads) {
            let mentions = 0;
            let unread = false;
            for (const myChannel of myChannels) {
                mentions += myChannel.mentionsCount;
                unread = unread || myChannel.isUnread;
            }

            unreads.mentions = mentions;
            unreads.unread = unread;
            subscriptions.set(serverUrl, unreads);
            updateTotal();
        }
    };

    const serversObserver = async (servers: ServersModel[]) => {
        registeredServers.current = sortServers(servers, intl);

        // unsubscribe mentions from servers that were removed
        const allUrls = new Set(servers.map((s) => s.url));
        const subscriptionsToRemove = [...subscriptions].filter(([key]) => !allUrls.has(key));
        for (const [key, map] of subscriptionsToRemove) {
            map.subscription?.unsubscribe();
            subscriptions.delete(key);
            updateTotal();
        }

        for (const server of servers) {
            const {lastActiveAt, url} = server;
            if (lastActiveAt && url !== currentServerUrl && !subscriptions.has(url)) {
                const unreads: UnreadSubscription = {
                    mentions: 0,
                    unread: false,
                };
                subscriptions.set(url, unreads);
                unreads.subscription = subscribeUnreadAndMentionsByServer(url, unreadsSubscription);
            } else if ((!lastActiveAt || url === currentServerUrl) && subscriptions.has(url)) {
                subscriptions.get(url)?.subscription?.unsubscribe();
                subscriptions.delete(url);
                updateTotal();
            }
        }
    };

    const onPress = useCallback(() => {
        if (registeredServers.current?.length) {
            const renderContent = () => {
                return (
                    <ServerList servers={registeredServers.current!}/>
                );
            };

            const snapPoints = ['50%', 10];
            if (registeredServers.current.length > 3) {
                snapPoints[0] = '90%';
            }

            const closeButtonId = 'close-your-servers';
            bottomSheet({
                closeButtonId,
                renderContent,
                snapPoints,
                theme,
                title: intl.formatMessage({id: 'your.servers', defaultMessage: 'Your servers'}),
            });
        }
    }, [isTablet, theme]);

    useEffect(() => {
        const subscription = subscribeAllServers(serversObserver);

        return () => {
            subscription?.unsubscribe();
            subscriptions.forEach((unreads) => {
                unreads.subscription?.unsubscribe();
            });
            subscriptions.clear();
        };
    }, []);

    return (
        <ServerIcon
            hasUnreads={total.unread}
            mentionCount={total.mentions}
            onPress={onPress}
            style={styles.icon}
            testID={'channel_list.servers.server_icon'}
        />
    );
}

