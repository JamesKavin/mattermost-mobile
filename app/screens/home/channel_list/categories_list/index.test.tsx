// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Database from '@nozbe/watermelondb/Database';
import React from 'react';

import {SYSTEM_IDENTIFIERS} from '@constants/database';
import ServerDataOperator from '@database/operator/server_data_operator';
import {getTeamById} from '@queries/servers/team';
import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import CategoriesList from '.';

describe('components/categories_list', () => {
    let database: Database;
    let operator: ServerDataOperator;
    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
        operator = server.operator;

        const team = await getTeamById(database, TestHelper.basicTeam!.id);
        await database.write(async () => {
            await team?.update(() => {
                team.displayName = 'Test Team!';
            });
        });
    });

    it('should render', () => {
        const wrapper = renderWithEverything(
            <CategoriesList
                isTablet={false}
                teamsCount={1}
                channelsCount={1}
            />,
            {database},
        );
        expect(wrapper.toJSON()).toBeTruthy();
    });

    it('should render channel list with thread menu', () => {
        const wrapper = renderWithEverything(
            <CategoriesList
                isCRTEnabled={true}
                isTablet={false}
                teamsCount={1}
                channelsCount={1}
            />,
            {database},
        );
        expect(wrapper.toJSON()).toBeTruthy();
    });

    it('should render team error', async () => {
        await operator.handleSystem({
            systems: [{id: SYSTEM_IDENTIFIERS.CURRENT_TEAM_ID, value: ''}],
            prepareRecordsOnly: false,
        });

        const wrapper = renderWithEverything(
            <CategoriesList
                isTablet={false}
                teamsCount={0}
                channelsCount={1}
            />,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

        await operator.handleSystem({
            systems: [{id: SYSTEM_IDENTIFIERS.CURRENT_TEAM_ID, value: TestHelper.basicTeam!.id}],
            prepareRecordsOnly: false,
        });
    });

    it('should render channels error', () => {
        const wrapper = renderWithEverything(
            <CategoriesList
                isTablet={false}
                teamsCount={1}
                channelsCount={0}
            />,
            {database},
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
