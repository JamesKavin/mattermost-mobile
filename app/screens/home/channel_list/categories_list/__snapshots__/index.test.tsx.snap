// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/categories_list should render channels error 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "maxWidth": "100%",
      },
    }
  }
  collapsable={false}
  style={
    {
      "backgroundColor": "#1e325c",
      "flex": 1,
      "maxWidth": "100%",
      "paddingLeft": 18,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "marginLeft": 0,
        },
      }
    }
    collapsable={false}
    style={
      {
        "marginLeft": 0,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            {
              "color": "#ffffff",
              "fontFamily": "Metropolis-SemiBold",
              "fontSize": 28,
              "fontWeight": "600",
              "lineHeight": 36,
            }
          }
          testID="channel_list_header.team_display_name"
        >
          Test Team!
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          {
            "bottom": 30,
            "left": 20,
            "right": 20,
            "top": 10,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.08)",
            "borderRadius": 14,
            "height": 28,
            "justifyContent": "center",
            "opacity": 1,
            "width": 28,
          }
        }
        testID="channel_list_header.plus.button"
      >
        <Icon
          name="plus"
          style={
            {
              "color": "rgba(255,255,255,0.8)",
              "fontSize": 18,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "paddingRight": 60,
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          {
            "color": "rgba(255,255,255,0.64)",
            "fontFamily": "OpenSans-SemiBold",
            "fontSize": 11,
            "fontWeight": "600",
            "lineHeight": 16,
          }
        }
        testID="channel_list_header.server_display_name"
      />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.08)",
          "borderRadius": 60,
          "height": 120,
          "justifyContent": "center",
          "width": 120,
        }
      }
    >
      <Icon
        name="alert-circle-outline"
        style={
          {
            "color": "rgba(255,255,255,0.48)",
            "fontSize": 72,
            "lineHeight": 72,
          }
        }
      />
    </View>
    <Text
      style={
        [
          {
            "fontFamily": "Metropolis-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 28,
          },
          {
            "color": "#ffffff",
            "marginTop": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      Couldn't load Test Team!
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          {
            "color": "#ffffff",
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      There was a problem loading content for this team.
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "flex": 0,
          "height": 48,
          "justifyContent": "center",
          "marginTop": 24,
          "opacity": 1,
          "paddingHorizontal": 24,
          "paddingVertical": 14,
        }
      }
    >
      <Text
        style={
          [
            {
              "alignItems": "center",
              "fontFamily": "OpenSans-SemiBold",
              "fontWeight": "600",
              "justifyContent": "center",
              "padding": 1,
              "textAlignVertical": "center",
            },
            {
              "fontSize": 16,
              "lineHeight": 18,
              "marginTop": 1,
            },
            {
              "color": "#1c58d9",
            },
          ]
        }
      >
        Retry
      </Text>
    </View>
  </View>
</View>
`;

exports[`components/categories_list should render team error 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "maxWidth": "100%",
      },
    }
  }
  collapsable={false}
  style={
    {
      "backgroundColor": "#1e325c",
      "flex": 1,
      "maxWidth": "100%",
      "paddingLeft": 18,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "marginLeft": 0,
        },
      }
    }
    collapsable={false}
    style={
      {
        "marginLeft": 0,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "space-between",
            },
            {
              "flex": 1,
            },
          ]
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            {
              "color": "rgba(255,255,255,0.64)",
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 14,
              "fontWeight": "600",
              "lineHeight": 20,
            }
          }
          testID="channel_list_header.team_display_name"
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="channel_list_header.logout.button"
      >
        <Text
          style={
            {
              "color": "rgba(255,255,255,0.64)",
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 14,
              "fontWeight": "600",
              "lineHeight": 20,
            }
          }
          testID="channel_list_header.team_display_name"
        >
          Log out
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      animatedStyle={
        {
          "value": {
            "marginRight": 8,
            "opacity": 1,
            "width": 40,
          },
        }
      }
      collapsable={false}
      style={
        {
          "marginRight": 8,
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255,255,255,0.12)",
                "borderRadius": 8,
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 20,
                "width": 40,
              },
              false,
            ]
          }
        >
          <Icon
            color="rgba(255,255,255,0.56)"
            name="filter-variant"
            size={24}
          />
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "rgba(255,255,255,0.12)",
          "borderRadius": 8,
          "flex": 1,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "flex-start",
          "marginVertical": 20,
          "padding": 8,
        }
      }
      testID="channel_list_subheader.search_field.button"
    >
      <Icon
        name="magnify"
        style={
          {
            "color": "rgba(255,255,255,0.72)",
            "fontSize": 24,
            "width": 24,
          }
        }
      />
      <Text
        style={
          {
            "color": "rgba(255,255,255,0.72)",
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
            "marginLeft": 5,
            "marginTop": 1,
          }
        }
      >
        Find channels...
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.08)",
          "borderRadius": 60,
          "height": 120,
          "justifyContent": "center",
          "width": 120,
        }
      }
    >
      <Icon
        name="alert-circle-outline"
        style={
          {
            "color": "rgba(255,255,255,0.48)",
            "fontSize": 72,
            "lineHeight": 72,
          }
        }
      />
    </View>
    <Text
      style={
        [
          {
            "fontFamily": "Metropolis-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 28,
          },
          {
            "color": "#ffffff",
            "marginTop": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      Couldn't load categories for 
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          {
            "color": "#ffffff",
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      There was a problem loading content for this server.
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "flex": 0,
          "height": 48,
          "justifyContent": "center",
          "marginTop": 24,
          "opacity": 1,
          "paddingHorizontal": 24,
          "paddingVertical": 14,
        }
      }
    >
      <Text
        style={
          [
            {
              "alignItems": "center",
              "fontFamily": "OpenSans-SemiBold",
              "fontWeight": "600",
              "justifyContent": "center",
              "padding": 1,
              "textAlignVertical": "center",
            },
            {
              "fontSize": 16,
              "lineHeight": 18,
              "marginTop": 1,
            },
            {
              "color": "#1c58d9",
            },
          ]
        }
      >
        Retry
      </Text>
    </View>
  </View>
</View>
`;
